param/parameterized.py

```diff
@@ -58,6 +60,51 @@
 VERBOSE = INFO - 1
 logging.addLevelName(VERBOSE, "VERBOSE")
 
+
+def _deprecate_positional_args(func):
+    """Internal decorator for methods that issues warnings for positional arguments
+
+    Using the keyword-only argument syntax in pep 3102, arguments after the
+    ``*`` will issue a warning when passed as a positional argument.
+
+    Adapted from scikit-learn
+    """
+    signature = inspect.signature(func)
+
+    pos_or_kw_args = []
+    kwonly_args = []
+    for name, param in signature.parameters.items():
+        if param.kind in (inspect.Parameter.POSITIONAL_OR_KEYWORD, inspect.Parameter.POSITIONAL_ONLY):
+            pos_or_kw_args.append(name)
+        elif param.kind == inspect.Parameter.KEYWORD_ONLY:
+            kwonly_args.append(name)
+            if param.default is inspect.Parameter.empty:
+                raise TypeError("Keyword-only param without default disallowed.")
+
+    @functools.wraps(func)
+    def inner(*args, **kwargs):
+        name = func.__qualname__.split('.')[0]
+        n_extra_args = len(args) - len(pos_or_kw_args)
+        if n_extra_args > 0:
+            extra_args = ", ".join(kwonly_args[:n_extra_args])
+
+            warnings.warn(
+                f"Passing '{extra_args}' as positional argument(s) to 'param.{name}' "
+                "was deprecated, please pass them as keyword arguments.",
+                DeprecationWarning,
+                stacklevel=2,
```

Comment:
I think it would be beneficial to have a custom warning here. This will make it possible for our other project to raise the warning as an error in our test suite. 

And have you checked if the `stacklevel` is correct?

Reply:
Yep `stacklevel` is correct, I ran the test suite and got correct warnings (lots of them!).

> I think it would be beneficial to have a custom warning here. This will make it possible for our other project to raise the warning as an error in our test suite.

Did you already figure out how to run a test suite and make it fail on some specific warnings only?

Reply:
> Yep stacklevel is correct, I ran the test suite and got correct warnings (lots of them!).

Sounds good, and this is also the case for packages outside Param? 

> Did you already figure out how to run a test suite and make it fail on some specific warnings only?

It is the same syntax as ignoring a file in `filterwarnings`. See [here](https://github.com/holoviz/holoviews/blob/main/pyproject.toml). But I can see that there is the possibility to reference the "path" with something like this `error::DeprecationWarning:param.parameterized`. 

Though, this will raise errors for all `DeprecationWarning`s. By having custom warnings, we will have finer control.

Reply:
> Sounds good, and this is also the case for packages outside Param?

Tested it in the context of other packages and it looked fine.



> It is the same syntax as ignoring a file in `filterwarnings`. See [here](https://github.com/holoviz/holoviews/blob/main/pyproject.toml). But I can see that there is the possibility to reference the "path" with something like this `error::DeprecationWarning:param.parameterized`.

Great, I've added a `ParamDeprecationWarning` class, being a subclass of `DeprecationWarning`. The test suite fails now if this warning is emitted. I noticed that unfortunately the *examples* tests (I have just added that in this PR) don't fail when this warning is emitted, I presume this has to do with `nbval` :/